api/app/references/api.py

```diff
@@ -105,3 +114,41 @@ def get(self):
 Kind regards,
 The {event_description} team.
 """
+
+
+class ReferenceAPI(ReferenceMixin, restful.Resource):
+
+    @marshal_with(reference_fields)
+    @auth_required
+    def get(self):
+        LOGGER.debug('Received get request for reference')
+        user = user_repository.get_by_id(g.current_user['id'])
+        if not (user or user.is_admin):
+            return FORBIDDEN
+
+        args = self.get_req_parser.parse_args()
+        response = response_repository.get_by_id_and_user_id(args['response_id'], user.id)
+        if not response:
+            return RESPONSE_NOT_FOUND
+
+        reference_responses = reference_request_repository.get_reference_by_response_id(response.id)
+        return [reference_response.Reference for reference_response in reference_responses], 200
+
+    def post(self):
+        LOGGER.debug('Received post request for reference')
+        args = self.post_req_parser.parse_args()
+        token = args['token']
+        uploaded_document = args['uploaded_document']
+        reference_request = reference_request_repository.get_by_token(token)
+        if not reference_request:
+            return REFRERENCE_REQUEST_WITH_TOKEN_NOT_FOUND
+
+        reference = reference_request_repository.get_reference_by_reference_request_id(reference_request.id)
+        if reference:
+            return DUPLICATE_REFERENCE_SUBMISSION
```

Comment:
With the relationships setup between ReferenceRequest and Reference, on reference_request you will be able to check if an item exists in `reference_request.references` and then return the duplicate error. Even better you can perhaps wrap this check in a method in the ReferenceRequest class called `has_reference()`